server/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -607,4 +602,41 @@ void unregisterChildNode(DiscoveryNode node) {
         }
     }
 
+    /**
+     * Start tracking a cancellable task with its tcp channel, so if the channel gets closed we can get a set of
+     * pending tasks associated that channel and cancel them as these results won't be retrieved by the parent task.
+     *
+     * @return a releasable that should be called when this pending task is completed
+     */
+    public Releasable startTrackingCancellableChannelTask(TcpChannel channel, CancellableTask task) {
+        assert cancellableTasks.containsKey(task.getId()) : "task [" + task.getId() + "] is not registered yet";
+        final ChannelPendingTaskTracker tracker = channelPendingTaskTrackers.computeIfAbsent(channel, k -> new ChannelPendingTaskTracker());
+        tracker.pendingTasks.add(task);
+        if (tracker.registered.compareAndSet(false, true)) {
+            channel.addCloseListener(ActionListener.wrap(
+                r -> {
+                    final Set<CancellableTask> pendingTasks = Collections.unmodifiableSet(tracker.pendingTasks);
+                    for (Consumer<Set<CancellableTask>> listener : onChannelCloseListeners) {
+                        listener.accept(pendingTasks);
+                    }
+                },
+                e -> {
+                    assert false : new AssertionError("must not be here", e);
+                }));
+        }
+        return () -> tracker.pendingTasks.remove(task);
```

Comment:
Can we assert it was removed?

Reply:
++. Adjusted in https://github.com/elastic/elasticsearch/pull/56620/commits/09a612eb067c2173fa060e3e2bea3bbedabfde8e.